### PR TITLE
Extract dispatch methods into Dispatcher

### PR DESCRIPTION
**Pull request by kourge**
The `Failable` namespace contains stateful methods that concern event registration and dispatching. Not only are they intermingled with other methods, but in that arrangement it's not a reusable piece of mechanic.

This change extracts both the dispatcher moving parts as well as its tests out into its own module. The decision not to re-export its constituents through `index.ts` is intentional: this module is for internal consumption.

### CODE REVIEW

File: src/dispatcher/index.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+export interface Handler<T> {
+  (a: T): any;
+}
```
Comment (adamduffy) [inline src/dispatcher/index.ts]: there should be a clever way to enforce
```
export interface Handler<T1, T2> {
  (a: T1): T2;
}```
allHandlers would have to keep a map of state to type, somehow, if it's worth it.

Comment (adamduffy) [inline src/dispatcher/index.ts]: since this is not related to your change it's not blocking.

Comment (kourge) [inline src/dispatcher/index.ts]: I've given this some (unfinished) thought, but the problem is roughly: `Handler` as a construct on its own is not very useful; inside `failable/dispatcher` we can replace it with `Function` and it'll work just as fine. The real challenge is the triple of methods `addListener` / `removeListener` / `dispatch` and how they interact with handlers. Take `failable/failable`, for example:
- `addListener(state: string, handler: Function): void` (raw type)
  - `addListener<T>(state: 'success', handler: (data: T) => any): void`
  - `addListener(state: 'failure', handler: (error: Error) => any): void`
  - `addListener(state: 'pending', handler: () => any): void`
- `removeListener(state: string, handler?: Function): void` (raw type)
  - Same as the three `addListener` overloads above, but with `handler` being optional
- `dispatch(state: string, data: any): void` (raw type)
  - `dispatch(state: 'success', data: T): void`
  - `dispatch(state: 'failure', data: Error): void`
  - `dispatch(state: 'pending', data: void): void`

This problem is reminiscent of [`typed-fsa`](https://github.com/UrbanDoor/typed-fsa), except instead of tying a payload type to a string / symbol type, we're tying a function type to a string type.
